Granular variables, parallel computing, cache saving and loading, detailed outputting

lb4418 opened the pull request:
Implemented the following features:
* Granular variables
* Parallel function evaluations
* Cache cost saving and loading
* Detailed outputting of iterations

lb4418 committed:
Implement Cache saving and loading

lb4418 committed:
Add jld2 datasets and docstrings.

lb4418 committed:
Fix timing, add runtime and function eval limits

lb4418 committed:
Detailed outputting and cache hits.

lb4418 committed:
Implement granular variables and dynamic scaling

lb4418 committed:
Anisotropic OrthoMADS. Removed float errors.

lb4418 committed:
Merge branch 'cache-saving-loading' into lb

lb4418 committed:
Implement thread-safe parallel evaluations

lb4418 committed:
Add Mesh anisotropy

lb4418 committed:
Clean up of granular variables implementation.

lb4418 committed:
Merge branch 'cache-saving-loading' of https://github.com/lb4418/DirectSearch.jl into cache-saving-loading

lb4418 committed:
Merge branch 'cache-saving-loading' into lb

lb4418 committed:
Fix mesh size increase

lb4418 committed:
Merge branch 'lb' into parallel-evaluations

lb4418 committed:
Fix non-thread-safe constraint functions

lb4418 committed:
Merge branch 'parallel-evaluations' of https://github.com/lb4418/DirectSearch.jl into parallel-evaluations

lb4418 committed:
Change poll initilization bound

lb4418 committed:
Bug fix, and cost rounding

lb4418 committed:
Typo and SetMaxEvals fix

lb4418 committed:
Merge branch 'parallel-evaluations' into lb

lb4418 committed:
Add project documentation

lb4418 committed:
Add project documentation

lb4418 committed:
Doc fixes

lb4418 committed:
Merge branch 'lb-docs' of https://github.com/lb4418/DirectSearch.jl into lb-docs

lb4418 committed:
Update README and docs

lb4418 committed:
Add SetFullOutput function

lb4418 committed:
Remove BEng project specific details

lb4418 committed:
Merge branch 'lb' of https://github.com/lb4418/DirectSearch.jl into lb

imciner2 commented on README.md:
```diff
@@ -1,18 +1,17 @@
+<!-- [![Dev](https://img.shields.io/badge/docs-dev-blue.svg)](https://imperialcollegelondon.github.io/DirectSearch.jl/dev)
 ![](https://github.com/ImperialCollegeLondon/DirectSearch.jl/workflows/CI/badge.svg)
-[![codecov](https://codecov.io/gh/ImperialCollegeLondon/DirectSearch.jl/branch/master/graph/badge.svg)](https://codecov.io/gh/ImperialCollegeLondon/DirectSearch.jl)
+[![codecov](https://codecov.io/gh/ImperialCollegeLondon/DirectSearch.jl/branch/master/graph/badge.svg)](https://codecov.io/gh/ImperialCollegeLondon/DirectSearch.jl) -->
```
Why do we need these lines commented out?

imciner2 commented on docs/src/man/addstoppingconditions.md:
```diff
@@ -0,0 +1,39 @@
+# Adding Stopping Conditions
+
+The software runs the optimization process until one of the stopping conditions. There are five stopping conditions, which may be used by the users, or they can define their own.
```
```suggestion
The software runs the optimization process until one of the stopping conditions is met. There are five stopping conditions defined in the package, and the ability for custom stopping conditions to be defined.
```

imciner2 commented on docs/src/man/addstoppingconditions.md:
````diff
@@ -0,0 +1,39 @@
+(CheckStoppingCondition(p::DSProblem, s::T) where T <: AbstractStoppingCondition)::Bool
+```
+
+`StoppingConditionStatus` defines the textual representation of the stopping condition, which is used in reporting. It takes as argument the instance of the concrete stopping condition type and should return a `string`. If this function is ommited, then the stopping condition status will be represented as "Unknown stopping condition status".
````
```suggestion
`StoppingConditionStatus` defines the textual representation of the stopping condition, which is used in reporting. It takes as an argument the instance of the concrete stopping condition type and should return a `string`. If this function is ommited, then the stopping condition status will be represented as "Unknown stopping condition status".
```

imciner2 commented on docs/src/man/addstoppingconditions.md:
```diff
@@ -0,0 +1,39 @@
+
+## Native
+
+There are currently five stopping conditions that may be used:
```
```suggestion
There are currently five stopping conditions implemented in the package:
```

imciner2 commented on docs/src/man/cache.md:
```diff
@@ -0,0 +1,28 @@
+
+## Cache Saving and Loading
+
+The evaluated costs during the optimization process are stored in the cache as a dictionary with the point as the key, and its evaluted cost as the value. The cache may be saved in a file after running the optimization, and can be pre-loaded in order to reduce the number of function evaluations.
```
```suggestion
The evaluated costs during the optimization process are stored in the cache as a dictionary with the poll point as the key, and its evaluted cost as the value. The cache may be saved in a file after running the optimization, and can then be pre-loaded in order to reduce the number of function evaluations in a future run.
```

imciner2 commented on docs/src/man/usage.md:
````diff
@@ -23,46 +33,62 @@ SetObjective(p,obj)
 ```
-This will run MADS until either the iteration limit (default 1000), or precision limit (`Float64` precision) are reached. The reason for stopping can be accessed as the `status` variable within the problem.
+
+The same operation can be applied to all variables with `SetGranularities` (example for N=3):
````
```suggestion
The same operation can be applied to all variables with `SetGranularities` (example for a problem with 3 variables):
```

imciner2 commented on docs/src/man/usage.md:
````diff
@@ -23,46 +33,62 @@ SetObjective(p,obj)
+Optimize!(p)
 ```
-*(Functions for accessing this data will be added in future.)*
+This will run MADS until one of the defined stopping conditions is met. By default, the stopping conditions are set to the iteration limit (default 1000), function evaluation limit (default 5000), mesh precision limit (`Float64` precision) and poll precision limit (`Float64` precision). For more details on stopping conditions, and how to add a custom one see [Adding Stopping Conditions](@ref).
````
```suggestion
This will run MADS until one of the defined stopping conditions is met. For more details on stopping conditions, and how to add a custom one see [Adding Stopping Conditions](@ref).
```

imciner2 commented on docs/src/man/usage.md:
````diff
@@ -23,46 +33,62 @@ SetObjective(p,obj)
-```julia
-BumpIterationLimit(p) #Increments iteration limit by 100 iterations
-BumpIterationLimit(p, n) #Increments iteration limit by n iterations
+After optimization is finished, the detailed results are printed as in the following example:
````
```suggestion
After the optimization is finished, detailed results are printed as in the following example:
```

imciner2 commented on docs/src/man/usage.md:
```diff
@@ -138,10 +172,9 @@ As with the search step, it is set in `DSProblem`:
-Two poll steps are included. The first is LTMADS, which generates a set of directions from a basis generated from a semi-random lower triangular matrix. The other is OrthoMADS, a later algorithm that generates an orthogonal set of directions. OrthoMADS needs the dimension of the problem to be provided. By default, LTMADS is used.
-
-Note that OrthoMADS is deterministic, therefore using DirectSearch with OrthoMADS will always give the same result (assuming the objective function and constraints are also deterministic). LTMADS has a random element, and will therefore give different results every time it is run. For this reason, LTMADS may need several runs to achieve its best result.
+Two poll steps are included. The first is LTMADS, which generates a set of directions from a basis generated from a semi-random lower triangular matrix. The other is OrthoMADS, a later algorithm that generates an orthogonal set of directions. It was recently adapted to granular variables as in 2019 C. Audet, S. Le Digabel, and C. Tribes, but the same name is continued to be used. By default, LTMA ...
```
```suggestion
Two poll types are included. The first is LTMADS, which generates a set of directions from a basis generated from a semi-random lower triangular matrix. The other is OrthoMADS, a later algorithm that generates an orthogonal set of directions. It was recently adapted to granular variables in 2019 C. Audet, S. Le Digabel, and C. Tribes, but the same name is continued to be used. By default, LTMADS is used.
```

imciner2 commented on src/Constraints.jl:
```diff
@@ -247,7 +247,15 @@ end
-function ConstraintCachePush(c::Constraints{T}, x::Vector{T}, i::Int, h::T) where T
+function ConstraintCachePush(p::AbstractProblem, x::Vector{T}, i::Int, h::T) where T
+    if p.config.max_simultaneous_evaluations > 1
+        lock(() -> constraint_cache_push(p.constraints, x, i, h), p.config.parallel_lock)
```
Why do we need to use the same lock as the cost cache here? It would probably be better if we used a new lock for constraints so that we don't have contention between threads at different stages of the poll point evaluation.

imciner2 replied:
I have just split the constraint cache and the cost cache locks apart since they operate on different underlying storage items.

imciner2 commented on src/Core.jl:
```diff
@@ -272,6 +368,19 @@ Determine whether the set of trial points result in a dominating, improving, or
+end
+
+"""
+    valuatePointSequential!(p::DSProblem{FT}, trial_points::Vector{Vector{FT}})::IterationOutcome where {FT<:AbstractFloat}
```
```suggestion
    EvaluatePointSequential!(p::DSProblem{FT}, trial_points::Vector{Vector{FT}})::IterationOutcome where {FT<:AbstractFloat}
```

imciner2 commented on src/Core.jl:
```diff
@@ -347,64 +466,153 @@ function EvaluatePoint!(p::DSProblem{FT}, trial_points::Vector{Vector{FT}})::Ite
+                             trial_point::Vector{T})::Tuple{T, Bool} where T
+    if CacheQuery(p, trial_point)
+        p.status.cache_hits += 1
+        return (CacheGet(p, trial_point), true)
```
Why do we need to return if the point was in the cache here?

lb4418 replied:
We need its cost, and the boolean is for outputting that the point was found in the cache.

imciner2 commented on src/OrthoMADS.jl:
```diff
@@ -8,151 +7,42 @@ export OrthoMADS
-
-function GenerateDirections(N::Int64, o::OrthoMADS{T})::Matrix{T} where T
-    o.init_run || init_orthomads(N, o)
-    H = GenerateOMBasis(N, o.t, o.l)
```
I think it would be better to keep this version of OrthoMADS and add your new one as a new polling type so that we can continue to use the original OrthoMADS as a comparison point.

imciner2 commented on src/StoppingCondition.jl:
```diff
@@ -92,14 +103,189 @@ function BumpIterationLimit(p::DSProblem, i::Int)
+
+(GetMeshSizeVector(p::DSProblem{T})::Vector{T}) where T = p.config.mesh.δ
+
+get_min_mesh_size(::Type{Float64}) = 1.1102230246251565e-16
```
Where does this constant come from?

EdwardStables replied:
This originally defined in `min_mesh_size` which has been removed in this pull request. The constant was set as half the machine epsilon of a `Float64` (smallest increment between 1 and the next number represented by the type, described by the `eps` function). This was intended to define the smallest value a Float64 can represent, and therefore the minimum value for the mesh size.

However as floating points don't actually work in a linear fashion where this approach would work, I'd suggest changing this to a larger number that can be well represented by a `Float64` over a larger range. This was an oversight in the original code.

imciner2 commented on src/StoppingCondition.jl:
```diff
@@ -92,14 +103,189 @@ function BumpIterationLimit(p::DSProblem, i::Int)
+
+(GetPollSizeVector(p::DSProblem{T})::Vector{T}) where T = p.config.mesh.Δ
+
+get_min_poll_size(::Type{Float64}) = 1.1102230246251565e-16
```
Also, where does this constant come from?

imciner2 commented on src/Types.jl:
```diff
@@ -58,17 +62,20 @@ mutable struct Config{FT<:AbstractFloat, MT<:AbstractMesh, ST<:AbstractSearch, P
-    max_simultanious_evaluations::Int
+    num_threads::Int
+    max_simultaneous_evaluations::Int
+    parallel_lock::Threads.AbstractLock
```
Can we make this a concrete type? Having abstract types in code can cause performance problems when we constantly call into functions using them (like the locking/unlocking functions).

imciner2 replied:
I have just done this and made them `ReentrantLock` concrete types.

imciner2 commented:
I've left some comments, but overall thanks for doing this! I am slightly concerned by the lack of tests for the code, but I don't think you have time to do those right now. Please take a look at my comments and let me know if anything doesn't make sense.

EdwardStables commented:
Thanks for working on this! Please do try and have all test passing before merging :).

It looks like it's only the OrthoMADS checks that are failing, either keep the existing OrthoMADS code and rename your additions (as Ian suggests in https://github.com/ImperialCollegeLondon/DirectSearch.jl/pull/5/files#r665389289), or modify the tests to work with your additions.

lb4418 committed:
Fix unit tests and docs

imciner2 committed:
Split locks for cost and constraints

Split the locks since they operate on different underlying cache
objects. Also make the locks a concrete type since they are in the
critical path so we want type stability.

imciner2 commented:
Superseded by #7
